### ISSUE TITLE
AttributeError: 'unicode' object has no attribute 'rpartition'

### ISSUE THREAD

**Issue by jonathanunderwood** [sshuttle/sshuttle#87]
If the remote end doesn't have python 2.7, sshuttle will bail with:

Traceback (most recent call last):
  File "<string>", line 1, in ?
  File "assembler.py", line 18, in ?
AttributeError: 'unicode' object has no attribute 'rpartition'
client: fatal: server died with error code 1

Ideally, sshuttle would support python >=2.4 on the remote end in order to support RHEL5 and RHEL6 remotely. Failing that, a more user friendly error message informing the user that the remote end python is too out of date would be nice.

A downstream bug report is [here](https://bugzilla.redhat.com/show_bug.cgi?id=1321580)

**Comment by jonathanunderwood** [comment]
PR #81 is progress towards fixing this issue.

**Comment by vieira** [comment]
I opened #37 with some fixes to make it work but it broke some tests and there wasn't much interest at the time. I thought I was the only remaining Python 2.4 user in the world. :stuck_out_tongue_winking_eye: 

I rebased sometime in January and opened #56 and @brianmay was kind enough to fix most of the tests in #81 even though I think he is not affected by this nor very excited about maintaining Python 2.4 code. 

#83 should have fixed the last few problems and one or two tests. I have also tested manually against a range of Python versions from 2.4 to 3.5 and it seems to be working nicely. 

Do you still have any problems with #83?

**Comment by jonathanunderwood** [comment]
I can understand the lack of excitement about maintaining python 2.4 code :). I wonder if a half way compromise might be to support python 2.6 - that covers rhel6 era distros, and tools such as six.py and pasteurize make it fairly painless.

Is the only blocker on merging #81 the fixing of the tests? If so, when I am back from travels, I could have a look at pitching in there.

**Comment by vieira** [comment]
AFAIK in #83 (which includes all commits from #81 plus some extra fixes) all tests are OK. It should work with all Python versions from 2.4 to 3.5. @brianmay already fixed the tests which I think was one of the blockers. He might have other considerations/reasons for not wanting to merge, I'm not sure.

Although I have tested #83 against Python 2.4-3.5 on a Debian server, if you can also test and report any problems it would be nice (clone vieira/sshuttle, checkout old_python).

**Comment by jonathanunderwood** [comment]
OK, I have taken current git from Brian and added the vieira/old_python commits on top and built a new Fedora package for testing, so hopefully that'll help shake out any problems. In case you're interested, the build is here:

http://koji.fedoraproject.org/koji/taskinfo?taskID=13540955

**Comment by racker-tyler** [comment]
I also noticed this issue when installing it fresh from brew for mac. It appears that the server I am connecting to is using python2.4 :( I doubt that itll get updated anytime soon either. Would I need to reach out to the owners of the homebrew repo's to copy the commits and build a new repo or is this going to be incorporated into a newer version?

**Comment by brianmay** [comment]
Released version 0.78.0; closing.

**Comment by alphazo** [comment]
Hello, sshuttle 0.78.0 doesn't work (with one of my Debian server since I have upgraded it. I'm getting a UnicodeDecodeError even thought it's not the exact same one as the one reported in this bug.
On the server side, I have a Debian Wheezy 7.1 using openssh-server 6.6p1-4~bpo70+1 and Python 2.7.3-6+deb7u2.
On the client side, I'm running Archlinux with sshuttle 0.78.0, openssh 7.2p2-1, Python 3.5.1-2 (default python) but Python2 2.7.11-3 is also installed.

When launching sshuttle I'm getting:
```
sshuttle --dns -vvv -r dedibox 0/0
Starting sshuttle proxy.
firewall manager: Starting firewall with Python version 3.5.1
firewall manager: ready method name nat.
IPv6 enabled: False
UDP enabled: False
DNS enabled: True
Binding redirector: 12300
TCP redirector listening on ('127.0.0.1', 12300).
TCP redirector listening with <socket.socket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 12300)>.
Binding DNS: 12300
DNS listening on ('127.0.0.1', 12300).
DNS listening with <socket.socket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_DGRAM, proto=0, laddr=('127.0.0.1', 12300)>.
Starting client with Python version 3.5.1
c : connecting to server...
c : executing: ['ssh', 'dedibox', '--', 'exec /bin/sh -c \'P=python3.5; $P -V 2>/dev/null || P=python; exec "$P" -c \'"\'"\'import sys, os; verbosity=3; sys.stdin = os.fdopen(0, "rb"); exec(compile(sys.stdin.read(958), "assembler.py", "exec"))\'"\'"\'\'']
c :  > channel=0 cmd=PING len=7 (fullness=0)
server: assembling u'sshuttle' (7 bytes)
server: assembling u'sshuttle.cmdline_options' (27 bytes)
server: assembling u'sshuttle.helpers' (949 bytes)
server: assembling u'sshuttle.ssnet' (5540 bytes)
server: assembling u'sshuttle.hostwatch' (2361 bytes)
server: assembling u'sshuttle.server' (3091 bytes)
Starting server with Python version 2.7.3
 s: latency control setting = True
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "assembler.py", line 37, in <module>
  File "sshuttle.server", line 236, in main
  File "sshuttle.server", line 87, in list_routes
  File "sshuttle.server", line 69, in _list_routes
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 63: ordinal not in range(128)
c : fatal: server died with error code 1
```
